argon2_remote_support.go

```diff
@@ -0,0 +1,460 @@
+// ocurred.
+type Argon2RemoteError struct {
+	ErrorType   Argon2RemoteErrorType
+	ErrorString string
+}
+
+func (e *Argon2RemoteError) Error() string {
+	str := new(bytes.Buffer)
+	fmt.Fprintf(str, "cannot process KDF request: %v", e.ErrorType)
+	if e.ErrorString != "" {
+		fmt.Fprintf(str, " (%s)", e.ErrorString)
+	}
+	return str.String()
+}
+
+func (o *Argon2RemoteOutput) Err() error {
+	if o.ErrorType == "" {
+		return nil
+	}
+	return &Argon2RemoteError{
+		ErrorType:   o.ErrorType,
+		ErrorString: o.ErrorString,
+	}
+}
+
+const (
+	notArgon2RemoteProcess     uint32 = 0
+	readyArgon2RemoteProcess   uint32 = 1
+	expiredArgon2RemoteProcess uint32 = 2
+)
+
+var (
+	argon2RemoteProcessStatus uint32 = notArgon2RemoteProcess
+)
+
+// SetIsArgon2RemoteProcess marks this process as being a remote processs capable of running
+// Argon2 in process. After calling this, [InProcessArgon2KDF] will return a real implementation
+// that runs in process.
+func SetIsArgon2RemoteProcess() {
+	if !atomic.CompareAndSwapUint32(&argon2RemoteProcessStatus, notArgon2RemoteProcess, readyArgon2RemoteProcess) {
+		panic("cannot call SetIsArgon2RemoteProcess more than once")
+	}
+}
+
+// RunArgon2RequestInRemoteProcess runs the specified argon2 request, and returns a response. This
+// function can only be called once in a process. Subsequent calls in the same process will result
+// in an error response being returned.
+func RunArgon2RequestInRemoteProcess(input *Argon2RemoteInput) *Argon2RemoteOutput {
+	if !atomic.CompareAndSwapUint32(&argon2RemoteProcessStatus, readyArgon2RemoteProcess, expiredArgon2RemoteProcess) {
+		switch atomic.LoadUint32(&argon2RemoteProcessStatus) {
+		case expiredArgon2RemoteProcess:
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorConsumedProcess,
+				ErrorString: "cannot run more than once in the same process",
+			}
+		default:
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorProcessNotConfigured,
+				ErrorString: "cannot run in a process that isn't configured as an Argon2 remote process",
+			}
+		}
+	}
+
+	switch input.Mode {
+	case Argon2id, Argon2i:
+		// ok
+	default:
+		return &Argon2RemoteOutput{
+			Command:     input.Command,
+			ErrorType:   Argon2RemoteErrorInvalidMode,
+			ErrorString: fmt.Sprintf("invalid mode: %q", string(input.Mode)),
+		}
+	}
+
+	costParams := &Argon2CostParams{
+		Time:      input.Time,
+		MemoryKiB: input.MemoryKiB,
+		Threads:   input.Threads,
+	}
+	if costParams.Time == 0 {
+		return &Argon2RemoteOutput{
+			Command:     input.Command,
+			ErrorType:   Argon2RemoteErrorInvalidTimeCost,
+			ErrorString: "invalid time cost: cannot be zero",
+		}
+	}
+	if costParams.Threads == 0 {
+		return &Argon2RemoteOutput{
+			Command:     input.Command,
+			ErrorType:   Argon2RemoteErrorInvalidThreads,
+			ErrorString: "invalid threads: cannot be zero",
+		}
+	}
+
+	switch input.Command {
```

Comment:
maybe a bit more of the error cases below should be tested

Reply:
I've added some more test cases.